torchvision/prototype/models/resnet.py

```diff
@@ -41,6 +41,16 @@ def _resnet(
     **kwargs: Any,
 ) -> ResNet:
     if weights is not None:
+        # I remember Francisco raising concern about overriding num_class.
+        # IMHO this is fine, I would consider this to be a case of
+        # "we have param A and param B, param B's default depends on the value of A."
+        # which is a very common pattern.
+        # However, in order to prevent users to do wrong things, I think we could raise
+        # an error if the user specified an incorrect 'num_classes':
```

Comment:
In the prototype there are 3 different overwrites similar to this. (all other overwrites are pre-existing)

The 1st is for the `num_classes` key of `kwargs` as in this case. This applies to all classification models.

The 2nd is for the `backend` key of `kwargs` (the case is identical to previous). This applies to all the quantized models like this:
https://github.com/pytorch/vision/blob/272e080c46ebef2b74f89c9cc94d84a9dc541bde/torchvision/prototype/models/quantization/mobilenetv3.py#L38

The 3rd is for the `num_classes` parameter of detection/segmentation models:
https://github.com/pytorch/vision/blob/272e080c46ebef2b74f89c9cc94d84a9dc541bde/torchvision/prototype/models/detection/faster_rcnn.py#L94

The latter case is a bit of an oddity as highlighted earlier by Nicolas because that's a parameter defined for Detection and Segmentation but not Classification. As you understand, when it's a parameter and not a `kwargs` value, we can't tell if it was the user who passed this or if they relied on the default value of the method and they forgot to pass the right value. So it's unclear how to implement the proposal below in all cases.

Perhaps given we already do these silent overwrites in a bunch of places (see GoogleNet, InceptionV3, ResNeXt etc), we don't have to throw a warning/error. Thoughts @NicolasHug @fmassa ?

Reply:
> when it's a parameter and not a kwargs value, we can't tell if it was the user who passed this or if they relied on the default value of the method and they forgot to pass the right value

That's a good point.

My suggestion then would be to change the default of `num_classes=91` (I don't remember the exact number) to `num_classes='auto'`, which would:
- defaults to 91 if weights is None - this preserves BC
- defaults to `len(weights.meta["categories"])` if  weights is passed

I think this is backward compatible, and will allow us to safely raise an error when the user manually overwrites `num_classes` to a wrong value.

Reply:
Yes, this idea is possible and BC. 

Instead of `'auto'` we could use `None` as this is a common idiom used in TorchVision. Unfortunately, this idiom has caused complaints (see #4905) among users. Some extra handling will be necessary to achieve full BC, for example if num_classes=None and weights=None then it should become `91` but all these are possible.

I'll wait for @fmassa to comment on the pending issues before tackling this as the change is big and in multiple places.

Reply:
The default value of the `num_classes` changed to `None`. So this is partially addressed at commit f253f4c104bde286891dacdd0bbd7c87b9e74535 of PR #4948

Reply:
Ok with me to let the default value of `num_classes` to be `None`

Reply:
The silent overwrites of `num_classes` and other similar config params is addressed at PR #4948 commit 9ff811ba6ae4af939aa8fb6f44b0e7b5eaef818a